File: packages/agentspec-core/src/parser.ts
/**
 * Parser module — Parse YAML/JSON/auto-detect into AgentSpecFile
 */

import yaml from 'js-yaml';
import type { AgentSpecFile } from './types';

export type FormatType = 'json' | 'yaml' | 'xml';

export function parseJSON(input: string): AgentSpecFile {
  try {
    return JSON.parse(input) as AgentSpecFile;
  } catch (e) {
    throw new Error(`Failed to parse JSON: ${(e as Error).message}`);
  }
}

export function parseYAML(input: string): AgentSpecFile {
  try {
    const result = yaml.load(input);
    if (typeof result !== 'object' || result === null) {
      throw new Error('YAML did not produce an object');
    }
    return result as AgentSpecFile;
  } catch (e) {
    throw new Error(`Failed to parse YAML: ${(e as Error).message}`);
  }
}

export function detectFormat(input: string): FormatType {
  const trimmed = input.trim();
  if (trimmed.startsWith('{') || trimmed.startsWith('[')) {
    return 'json';
  }
  if (trimmed.startsWith('<')) {
    return 'xml';
  }
  if (trimmed.includes(':')) {
    return 'yaml';
  }
  return 'json';
}

export function parseSpec(input: string, format?: FormatType): AgentSpecFile {
  const detectedFormat = format || detectFormat(input);

  switch (detectedFormat) {
    case 'json':
      return parseJSON(input);
    case 'yaml':
      return parseYAML(input);
    default:
      throw new Error(`Unsupported format: ${detectedFormat}`);
  }
}

File: packages/agentspec-py/agentspec/parser.py
"""AgentSpec Parser - Parse YAML/JSON into AgentSpecFile"""

from __future__ import annotations

import json
from typing import Literal, Optional

import yaml

from .types import AgentSpecFile

FormatType = Literal["json", "yaml", "xml"]


def parse_json(input_str: str) -> AgentSpecFile:
    try:
        data = json.loads(input_str)
        return AgentSpecFile(**data)
    except json.JSONDecodeError as e:
        raise ValueError(f"Failed to parse JSON: {e}")


def parse_yaml(input_str: str) -> AgentSpecFile:
    try:
        data = yaml.safe_load(input_str)
        if not isinstance(data, dict):
            raise ValueError("YAML did not produce an object")
        return AgentSpecFile(**data)
    except yaml.YAMLError as e:
        raise ValueError(f"Failed to parse YAML: {e}")


def detect_format(input_str: str) -> FormatType:
    trimmed = input_str.strip()
    if trimmed.startswith("{") or trimmed.startswith("["):
        return "json"
    if trimmed.startswith("<"):
        return "xml"
    if ":" in trimmed:
        return "yaml"
    return "json"


def parse_spec(input_str: str, format: Optional[FormatType] = None) -> AgentSpecFile:
    fmt = format or detect_format(input_str)
    if fmt == "json":
        return parse_json(input_str)
    elif fmt == "yaml":
        return parse_yaml(input_str)
    else:
        raise ValueError(f"Unsupported format: {fmt}")
